source/server/listener_manager_impl.cc

```diff
@@ -98,15 +120,20 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, ListenerManag
       (config.filter_chains().size() == 1 &&
        config.filter_chains()[0].filter_chain_match().sni_domains().empty());
 
+  std::string digest;
   uint32_t has_tls = 0;
   uint32_t has_stk = 0;
   for (const auto& filter_chain : config.filter_chains()) {
     std::vector<std::string> sni_domains(filter_chain.filter_chain_match().sni_domains().begin(),
                                          filter_chain.filter_chain_match().sni_domains().end());
-    if (filter_factories_.empty() || !filter_chain.filters().empty()) {
-      // Only one filter chain can have filters until multiple filter chains are properly supported.
-      RELEASE_ASSERT(filter_factories_.empty());
-      filter_factories_ = parent_.factory_.createFilterFactoryList(filter_chain.filters(), *this);
+    auto ret = parent_.factory_.createFilterFactoryList(filter_chain.filters(), *this);
```

Comment:
I don't think we need all of the hashing/interface changes here (and if we did need to hash we should use std::hash for this). Can't we literally just equality comparison the protos? (Basically if there are more than 1, just compare the rest to the first?

Reply:
At this point, filters are still opaque, and because we're using `google.protobuf.Struct` (i.e. JSON) for `config`, equal filter chains are not guaranteed to be equal in JSON format, because of anti-DoS protections for hash collisions, which randomize order of items in objects.

cc @htuch

Reply:
I'm not an expert on this, but:
1) I'm pretty sure some variant of https://github.com/google/protobuf/blob/master/src/google/protobuf/util/message_differencer.h would make this work correctly.

2) I think in order to get stable hashes you have to do: https://github.com/envoyproxy/envoy/blob/master/source/common/protobuf/utility.h#L74

Even if we have to do (2) and hash, can't we just do the simple thing and take the hash here in this function and then just compare subsequent? It would be close enough. I don't think we need to do the complicated pass between multiple functions stuff... Will defer to @htuch on this.

Reply:
OK, it seems that the `MessageUtil::hash()` works, I'll update the code to use it, thanks for pointing this out.